department-of-veterans-affairs/caseflow
# MattT/CASEFLOW-5464 Only show 'Return to X' task action if org X has closed their task.

Resolves [CASEFLOW-5464](https://vajira.max.gov/browse/CASEFLOW-5464)

### Description
```
Currently we are showing the Return to Education Service task action to BVA Intake 
users whenever the appeal is already assigned to an organization in the Education 
business line (EMO & RPOs). This PR aims to resolve that.
```

### Acceptance Criteria
- [ ] If a child `EducationDocumentSearchTask` is not closed (completed or cancelled) then the `Return to Education Service` task action is unavailable to BVA Intake users.
- [ ] If a child `VhaDocumentSearchTask` is not closed (completed or cancelled) then the `Return appeal to VHA` task action is unavailable to BVA Intake users.

### Testing Plan
1. Open the Rails console, and create an appeal with an `EducationDocumentSearchTask`:

```ruby
emo_task = FactoryBot.create(:education_document_search_task, :assigned)
appeal = emo_task.appeal
```
Your appeal's task `.treee` should appear as follows:
```ruby
                                        ┌────────────────────────────────────────────────────────────────────┐
Appeal 1669 (E 220502-1669 Original) ── │ ID   │ STATUS   │ ASGN_BY │ ASGN_TO      │ UPDATED_AT              │
└── RootTask                            │ 7136 │ on_hold  │         │ Bva          │ 2022-05-03 18:41:15 UTC │
    └── PreDocketTask                   │ 7137 │ on_hold  │         │ BvaIntake    │ 2022-05-03 18:41:15 UTC │
        └── EducationDocumentSearchTask │ 7138 │ assigned │         │ EducationEmo │ 2022-05-03 18:41:15 UTC │
                                        └────────────────────────────────────────────────────────────────────┘
```

2. Log into Caseflow as a BVA Intake user and navigate to the appeal's Queue page.
`http://localhost:3000/queue/appeals/<appeal.uuid>`

3. Click on the Actions dropdown and ensure that `Return to Education Service` is **not** present.

4. Complete the `EducationDocumentSearchTask`:
```ruby
emo_task.completed!
```

`appeal.treee`:

```ruby                                  ┌─────────────────────────────────────────────────────────────────────┐
Appeal 1669 (E 220502-1669 Original) ── │ ID   │ STATUS    │ ASGN_BY │ ASGN_TO      │ UPDATED_AT              │
└── RootTask                            │ 7136 │ on_hold   │         │ Bva          │ 2022-05-03 18:41:15 UTC │
    └── PreDocketTask                   │ 7137 │ assigned  │         │ BvaIntake    │ 2022-05-03 18:53:45 UTC │
        └── EducationDocumentSearchTask │ 7138 │ completed │         │ EducationEmo │ 2022-05-03 18:53:45 UTC │
                                        └─────────────────────────────────────────────────────────────────────┘
```

5. Refresh the page, select the dropdown once more, and make sure that this time the `Return to Education Service` does appear.

6. Add another `EducationDocumentSearchTask` to the appeal:
```ruby
emo_task2 = EducationDocumentSearchTask.create!(appeal: appeal, parent: emo_task.parent, assigned_at: Time.zone.now, assigned_to: EducationEmo.singleton)
```

`appeal.treee`:
```ruby                                   ┌─────────────────────────────────────────────────────────────────────┐
Appeal 1669 (E 220502-1669 Original) ── │ ID   │ STATUS    │ ASGN_BY │ ASGN_TO      │ UPDATED_AT              │
└── RootTask                            │ 7136 │ on_hold   │         │ Bva          │ 2022-05-03 18:41:15 UTC │
    └── PreDocketTask                   │ 7137 │ assigned  │         │ BvaIntake    │ 2022-05-03 18:53:45 UTC │
        ├── EducationDocumentSearchTask │ 7138 │ completed │         │ EducationEmo │ 2022-05-03 18:53:45 UTC │
        └── EducationDocumentSearchTask │ 7139 │ assigned  │         │ EducationEmo │ 2022-05-03 18:55:01 UTC │
                                        └─────────────────────────────────────────────────────────────────────┘
```

7. Refresh the appeal page again, and this time the 'Return to Education Service' task should not appear.

8. Add an `EducationAssessDocumentTask` to the appeal:

```ruby
rpo_task = EducationAssessDocumentationTask.create!(appeal: appeal, parent: emo_task2, assigned_at: Time.zone.now, assigned_to: EduRegionalProcessingOffice.first)
```

`appeal.treee`:
```ruby                                 ┌────────────────────────────────────────────────────────────────────────────────────┐
Appeal 1669 (E 220502-1669 Original) ─────────── │ ID   │ STATUS    │ ASGN_BY │ ASGN_TO                     │ UPDATED_AT              │
└── RootTask                                     │ 7136 │ on_hold   │         │ Bva                         │ 2022-05-03 18:41:15 UTC │
    └── PreDocketTask                            │ 7137 │ assigned  │         │ BvaIntake                   │ 2022-05-03 18:53:45 UTC │
        ├── EducationDocumentSearchTask          │ 7138 │ completed │         │ EducationEmo                │ 2022-05-03 18:53:45 UTC │
        └── EducationDocumentSearchTask          │ 7139 │ on_hold   │         │ EducationEmo                │ 2022-05-03 19:01:40 UTC │
            └── EducationAssessDocumentationTask │ 7140 │ assigned  │         │ EduRegionalProcessingOffice │ 2022-05-03 19:01:40 UTC │
                                                 └────────────────────────────────────────────────────────────────────────────────────┘
```

9. Refresh the page again and verify that the 'Return to Education Service' is still not there.

10. Complete the two latest tasks:
```ruby
rpo_task.completed!
emo_task2.completed!
```
`appeal.treee`:
```ruby                                 ┌────────────────────────────────────────────────────────────────────────────────────┐
Appeal 1669 (E 220502-1669 Original) ─────────── │ ID   │ STATUS    │ ASGN_BY │ ASGN_TO                     │ UPDATED_AT              │
└── RootTask                                     │ 7136 │ on_hold   │         │ Bva                         │ 2022-05-03 18:41:15 UTC │
    └── PreDocketTask                            │ 7137 │ assigned  │         │ BvaIntake                   │ 2022-05-03 19:03:20 UTC │
        ├── EducationDocumentSearchTask          │ 7138 │ completed │         │ EducationEmo                │ 2022-05-03 18:53:45 UTC │
        └── EducationDocumentSearchTask          │ 7139 │ completed │         │ EducationEmo                │ 2022-05-03 19:03:20 UTC │
            └── EducationAssessDocumentationTask │ 7140 │ completed │         │ EduRegionalProcessingOffice │ 2022-05-03 19:03:03 UTC │
                                                 └────────────────────────────────────────────────────────────────────────────────────┘
```

11. Refresh one last time and make sure that the 'Return to Education Service' task is once again visible in the actions dropdown.

12. Feel free to repeat these steps for the VHA business line and `VhaDocumentSearchTask`.
